socs/agents/hwp_gripper/agent.py

```diff
@@ -470,6 +525,87 @@ def run_and_append(func, *args, **kwargs):
                        check_shutdown=check_shutdown)
         time.sleep(1)
 
+        # We should stop schedule if we have an error in this task
+        if (not all(finished)) or any(aborted):
+            raise ValueError("Failed to grip HWP. Grippers are retracted.")
+
+        return data
+
+    def ungrip(self, session, params=None):
+        """ungrip()
+
+        **Task** - Series of commands to automatically ungrip the HWP.
+        This will return grippers to their home position, and retract
+        grippers as much as possible.
+        """
+
+        self._ungrip_hwp(session, check_shutdown=True)
```

Comment:
Rather than passing the `session` object to this helper function, it'd be better (and fairly straightforward) to set the `session.data` within the `ungrip()` function after returning `data` from `_ungrip_hwp(check_shutdown=True)`. So it would look like:
```
session.data = self._ungrip_hwp(check_shutdown=True)
```

This decouples `_ungrip_hwp()` a bit more, and you're already returning the `data` from that function.